File: #44_Wildcard_Matching/Solution.py
'''
Time O(m * n),
Space O(m * n) ->O(m)
1. if dp[i - 1][j - 1] is true and s[i] == p[j] or p[j] == '?' or p[j] == '*', dp[i][j] is true
2. if dp[i - 1][j] is true, and p[j] == '*' and s[i - 1] == s[i], then dp[i][j] is true
3. if dp[i][j - 1] is true, and p[j] == '*', then dp[i][j] is true
'''

class Solution:
	# @param {string} s
	# @param {string} p
	# @return {boolean}
	def isMatch(self, s, p):
		if not s and not p: # both empty
			return True
		if not p: # p is empty string
			return False
		if len(p) - p.count('*') > len(s):
			return False
		dp = [[False for j in range(len(p) + 1)] for i in range(len(s) + 1)]
		dp[0][0] = True
		for j in range(1, len(p) + 1):
			if dp[0][j - 1] and p[j - 1] == '*':
				dp[0][j] = True
		for i in range(1, len(s) + 1):
			for j in range(1, len(p) + 1):
				if dp[i - 1][j - 1] and (s[i - 1] == p[j - 1] or p[j - 1] == '?'):
					dp[i][j] = True
				elif dp[i - 1][j] and p[j - 1] == '*':
					dp[i][j] = True
				elif dp[i][j - 1] and p[j - 1] == '*':
					dp[i][j] = True
		return dp[len(s)][len(p)]

sol = Solution()
print sol.isMatch("abefcdgiescdfimde", "ab*cd?i*de")

File: #44_Wildcard_Matching/Solution.java
/* Greedy algorithm */
public class Solution {
	public boolean isMatch(String s, String p) {
		// if p.length is 0, return true only if s.length is 0 too
		if (p.length() == 0) return s.length() == 0;
		// index for s, index for p, check point for s, check point for p
		int is = 0, ip = 0, cs = -1, cp = -1;
		while(ip < p.length()) { // ip iterate chars in p
			// match
			if (is < s.length() && (s.charAt(is) == p.charAt(ip) || p.charAt(ip) == '?')) is++;
			// *, save check point
			else if (is < s.length() && p.charAt(ip) == '*') {
				cp = ip;
				cs = is;
			}
			// not match and no previous *
			else if (is < s.length() && cp < 0) return false;
			// not match but has previous *, return to it
			else if (is < s.length() && (cp >= 0)) {
				ip = cp;
				is = cs++;
			}
			else break; // if is reach the end of s, break loop
			ip++;
			if (is < s.length() && ip == p.length() && cp >= 0) {
				ip = cp + 1;
				is = cs++;
			}
		}
		// s is not over, p is over
		if (is < s.length() && ip == p.length()) return p.charAt(ip - 1) == '*';
		// s and p are both over
		if (is == s.length() && ip == p.length()) return true;
		// s is over but p is not over
		while(ip < p.length()) {
			if (p.charAt(ip) != '*') return false;
			ip++;
		}
		return true;
	}

	public static void main(String[] args) {
		Solution sol = new Solution();
		System.out.println(sol.isMatch("hi", "*?"));
	}
}